starknet/l2_minter/src/ethereum_minter.cairo

```diff
@@ -0,0 +1,252 @@
+        BookingClaimed: BookingClaimed,
+        BookingHandled: BookingHandled,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct Upgraded {
+        implementation: ClassHash
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct BookingHandled {
+        address: ContractAddress,
+        value: u256,
+        time: u64
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct BookingClaimed {
+        address: ContractAddress,
+        value: u256,
+    }
+
+    // Methods
+    #[constructor]
+    fn init(
+        ref self: ContractState,
+        projects_contract: ContractAddress,
+        slot: u256,
+        unit_price: u256,
+        max_supply: u256,
+        max_value_per_tx: u256,
+        min_value_per_tx: u256,
+    ) {
+        assert(!projects_contract.is_zero(), 'Projects contract cannot be 0');
+        assert(unit_price > 0, 'Unit price must be positive');
+        assert(max_supply > 0, 'Max supply must be positive');
+        assert(max_value_per_tx > 0, 'Max value per tx <= 0');
+        assert(min_value_per_tx > 0, 'Min value per tx <= 0');
+        assert(max_value_per_tx >= min_value_per_tx, 'Max < min per tx');
+        assert(max_supply >= max_value_per_tx, 'Max supply < max value per tx');
+
+        assert(slot != 0, 'Slot cannot be zero');
+
+        self._projects_contract.write(IERC3525Dispatcher { contract_address: projects_contract });
+        self._slot.write(slot);
+        self._unit_price.write(unit_price);
+        self._max_supply.write(max_supply);
+        self._max_value_per_tx.write(max_value_per_tx);
+        self._min_value_per_tx.write(min_value_per_tx);
+        self._current_max_supply.write(max_supply);
+    }
+
+    #[generate_trait]
+    #[external(v0)]
+    impl EthereumMinter of IEthereumMinter {
+        fn get_l1_minter_address(self: @ContractState) -> felt252 {
+            self._l1_minter_address.read()
+        }
+
+        fn sold_out(self: @ContractState) -> bool {
+            self._current_max_supply.read() < self._min_value_per_tx.read()
+        }
+
+        fn claim(ref self: ContractState, user_address: ContractAddress, id: u32) {
+            assert(self.sold_out(), 'Contract not sold out');
+
+            // [Check] Booking ok;
+            let mut booking = self._booked_values.read((user_address.into(), id));
+            assert(
+                booking.status == mint_status_to_u8(MintStatus::Booked(())), 'Booking not found'
+            );
+
+            let projects_contract = self._projects_contract.read();
+            let slot = self._slot.read();
+
+            // [Interaction] Mint
+            let token_id = self
+                ._projects_contract
+                .read()
+                .mintNew(user_address.into(), slot, booking.value);
+
+            // [Effect] Update Booking status
+            booking.status = mint_status_to_u8(MintStatus::Minted(()));
+            self._booked_values.write((user_address.into(), id), booking);
```

Comment:
Effect must be executed before Interaction I think

Reply:
done